io/v7/rfile/test/rfile.cxx

```diff
@@ -0,0 +1,568 @@
+      hist->FillRandom("gaus", 1000);
+      TFile file(fileGuard.GetPath().c_str(), "RECREATE");
+      auto *d = file.mkdir("a/b");
+      d->WriteObject(hist.get(), "hist");
+      d->WriteObject(hist.get(), "c/d");
+   }
+
+   {
+      auto file = RFile::OpenForReading(fileGuard.GetPath());
+      EXPECT_TRUE(file->Get<TH1D>("a/b/hist"));
+      // We won't find any object with a '/' in its name through RFile.
+      EXPECT_FALSE(file->Get<TH1D>("a/b/c/d"));
+   }
+}
+
+TEST(RFile, IterateKeys)
+{
+   FileRaii fileGuard("test_rfile_iteratekeys.root");
+
+   {
+      auto file = RFile::Recreate(fileGuard.GetPath());
+      TH1D a;
+      auto b = std::make_unique<TTree>();
+      std::string c = "0";
+      file->Put("a", a);
+      file->Put("b", *b);
+      file->Put("c", c);
+   }
+
+   {
+      auto file = RFile::OpenForReading(fileGuard.GetPath());
+      const auto expected = "a,b,c,";
+      std::string s = "";
+      for (const auto &key : file->GetKeys()) {
+         s += key.fName + ",";
+      }
+      EXPECT_EQ(expected, s);
+
+      // verify the expected iterator operations work
+      const auto expected2 = "b,c,";
+      s = "";
+      auto iterable = file->GetKeys();
+      auto it = iterable.begin();
+      std::advance(it, 1);
+      for (; it != iterable.end(); ++it) {
+         s += it->fName + ",";
+      }
+      EXPECT_EQ(expected2, s);
+   }
+}
+
+TEST(RFile, SaneHierarchy)
+{
+   // verify that we can't create weird hierarchies like:
+   //
+   // (root)
+   //   `--- "a/b": object
+   //   |
+   //   `--- "a": dir
+   //         |
+   //         `--- "b": object
+   //
+   // (who should "a/b" be in this case??)
+   //
+
+   FileRaii fileGuard("test_rfile_sane_hierarchy.root");
+
+   {
+      auto file = RFile::Recreate(fileGuard.GetPath());
+      std::string s;
+      file->Put("a", s);
+      EXPECT_THROW(file->Put("a/b", s), ROOT::RException);
+      file->Put("b/c", s);
+      file->Put("b/d", s);
+      EXPECT_THROW(file->Put("b/c/d", s), ROOT::RException);
+      EXPECT_THROW(file->Put("b", s), ROOT::RException);
+
+      EXPECT_NE(file->Get<std::string>("a"), nullptr);
+      EXPECT_EQ(file->Get<std::string>("a/b"), nullptr);
+      EXPECT_NE(file->Get<std::string>("b/c"), nullptr);
+      EXPECT_NE(file->Get<std::string>("b/d"), nullptr);
+      EXPECT_EQ(file->Get<std::string>("b/c/d"), nullptr);
+      EXPECT_EQ(file->Get<std::string>("b"), nullptr);
+   }
+}
+
+TEST(RFile, RefuseToCreateDirOverLeaf)
+{
+   FileRaii fileGuard("test_rfile_dir_over_leaf.root");
+   auto file = RFile::Recreate(fileGuard.GetPath());
+   std::string s;
+   file->Put("a/b", s);
+   try {
+      file->Put("a/b/c", s);
+      FAIL() << "creating a directory over a leaf path should fail.";
+   } catch (const ROOT::RException &ex) {
+      EXPECT_THAT(ex.what(), testing::HasSubstr("'a/b'"));
+      EXPECT_THAT(ex.what(), testing::HasSubstr("name already taken"));
+   }
+}
+
+TEST(RFile, IterateKeysRecursive)
```

Comment:
What does 'recursing' over the directories look like for 'complex' operation?  For example creating a graph representation of the file's content (text or in TBrowser nodes); Another example could be create Canvas and subCanvas display the histo of each directory (in an organization mimicking the directory organization).